src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -922,301 +767,124 @@ public SolutionState WithProjectDocumentsOrder(ProjectId projectId, ImmutableLis
-            var documentIdsByProjectId = documentIds.ToLookup(id => id.ProjectId);
-
-            var newSolutionState = this;
-
-            foreach (var documentIdsInProject in documentIdsByProjectId)
-            {
-                var oldProjectState = this.GetProjectState(documentIdsInProject.Key);
-
-                if (oldProjectState == null)
-                {
-                    throw new InvalidOperationException(string.Format(WorkspacesResources._0_is_not_part_of_the_workspace, documentIdsInProject.Key));
-                }
-
-                var removedDocumentStatesBuilder = ArrayBuilder<T>.GetInstance();
-
-                foreach (var documentId in documentIdsInProject)
-                {
-                    removedDocumentStatesBuilder.Add(getExistingTextDocumentState(oldProjectState, documentId));
-                }
-
-                var removedDocumentStatesForProject = removedDocumentStatesBuilder.ToImmutableAndFree();
-
-                var (newProjectState, compilationTranslationAction) = removeDocumentsFromProjectState(oldProjectState, documentIdsInProject.ToImmutableArray(), removedDocumentStatesForProject);
-
-                newSolutionState = newSolutionState.ForkProject(newProjectState,
-                    compilationTranslationAction,
-                    newFilePathToDocumentIdsMap: CreateFilePathToDocumentIdsMapWithRemovedDocuments(removedDocumentStatesForProject));
-            }
-
-            return newSolutionState;
-        }
-
-        /// <summary>
-        /// Creates a new solution instance that no longer includes the specified additional documents.
-        /// </summary>
-        public SolutionState RemoveAdditionalDocuments(ImmutableArray<DocumentId> documentIds)
-        {
-            return RemoveDocumentsFromMultipleProjects(documentIds,
-                (projectState, documentId) => projectState.AdditionalDocumentStates.GetRequiredState(documentId),
-                (projectState, documentIds, documentStates) => (projectState.RemoveAdditionalDocuments(documentIds), new CompilationAndGeneratorDriverTranslationAction.RemoveAdditionalDocumentsAction(documentStates)));
+            return ForkProject(oldProject, newProject);
         }
 
         /// <summary>
         /// Creates a new solution instance with the document specified updated to have the specified name.
         /// </summary>
-        public SolutionState WithDocumentName(DocumentId documentId, string name)
+        public StateChange WithDocumentName(DocumentId documentId, string name)
         {
             var oldDocument = GetRequiredDocumentState(documentId);
             if (oldDocument.Attributes.Name == name)
             {
-                return this;
+                var oldProject = GetRequiredProjectState(documentId.ProjectId);
+                return new(this, oldProject, oldProject);
             }
 
             return UpdateDocumentState(oldDocument.UpdateName(name), contentChanged: false);
```

Comment:
We absolutely should add more helpers for this, but I fully agree with your decision to leave those for follow-ups. :smile: